### PR TITLE
ENG-12733: Support multiple schema and procedure files

### PR DESCRIPTION
**Pull request by n3world**
Add support for providing multiple schema and jar files during the voltdb initialization

### CODE REVIEW

File: src/frontend/org/voltdb/compiler/VoltCompiler.java [hunk 1]
```diff
@@ -551,16 +551,19 @@ public boolean compileFromSchemaAndClasses(
             return false;
         }
 
-        InMemoryJarfile inMemoryUserJar = null;
+        InMemoryJarfile inMemoryUserJar = new InMemoryJarfile();
         ClassLoader originalClassLoader = m_classLoader;
         try {
-            if (classesJarPath != null && classesJarPath.exists()) {
+            m_classLoader = inMemoryUserJar.getLoader();
```
Comment (JadenW) [inline src/frontend/org/voltdb/compiler/VoltCompiler.java]: Why have to switch classLoader?

File: src/frontend/org/voltdb/VoltDB.java [hunk 2]
```diff
@@ -718,18 +718,25 @@ else if (arg.equalsIgnoreCase("buildstringoverride")) {
...
-                    }
-                    if (!m_stagedClassesPath.canRead()) {
-                        System.err.println("FATAL: Supplied classes jar file " + m_stagedClassesPath + " can't be read.");
-                        referToDocAndExit();
-                    }
-                    if (!m_stagedClassesPath.isFile()) {
-                        System.err.println("FATAL: Supplied classes jar file " + m_stagedClassesPath + " is not an ordinary file.");
-                        referToDocAndExit();
+                    for (String jarPath : Splitter.on(",").trimResults().omitEmptyStrings().split(args[++i])) {
+                        File staggedJar = new File(jarPath);
```
Comment (manjujames) [inline src/frontend/org/voltdb/VoltDB.java]: Nit: Should be stagedJar instead of staggedJar.

**Comment by manjujames** [comment]
Looks fine to me. I assume we have to test voltadmin tool manually?

Only question I have is about the user interface - Should it be possible for the user to specify a directory containing the schema/jar files? Might be a question for Ruth and Andrew. Could even be a separate ticket, if this change is to satisfy a specific customer request.

File: src/frontend/org/voltdb/compiler/VoltCompiler.java (hunk 1 above)
Comment (n3world) [reply src/frontend/org/voltdb/compiler/VoltCompiler.java]: The change was always done before. It appears that the class loader is used in methods called by compileInternal. Not sure of the reason why it is needed.

**Comment by n3world** [comment]
> Looks fine to me. I assume we have to test voltadmin tool manually?
> 
> Only question I have is about the user interface - Should it be possible for the user to specify a directory containing the schema/jar files? Might be a question for Ruth and Andrew. Could even be a separate ticket, if this change is to satisfy a specific customer request.

voltdb is partially tested by the voltdb-verbs-test which did find an issue with one of my earlier revisions. voltdb just blindly passes the value along to the java process so I would be surprised if there was an issue there.

Adding directory support wouldn't be hard but the question would be what do you take from the directory and is it a recursive search. For classes it could be all `.jar` files you find but for schema what extensions do you look for `.ddl` or `.sql`?

**Comment by manjujames** [comment]
> > Looks fine to me. I assume we have to test voltadmin tool manually?
> > Only question I have is about the user interface - Should it be possible for the user to specify a directory containing the schema/jar files? Might be a question for Ruth and Andrew. Could even be a separate ticket, if this change is to satisfy a specific customer request.
> 
> voltdb is partially tested by the voltdb-verbs-test which did find an issue with one of my earlier revisions. voltdb just blindly passes the value along to the java process so I would be surprised if there was an issue there.
> 
> Adding directory support wouldn't be hard but the question would be what do you take from the directory and is it a recursive search. For classes it could be all `.jar` files you find but for schema what extensions do you look for `.ddl` or `.sql`?

Yes, exactly. Which is why depending on how quickly Ruth needs it, it might be something to discuss with Andrew.

**Pull request merged**